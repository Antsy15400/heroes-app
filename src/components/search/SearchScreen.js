import { useMemo } from 'react'
import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";

import { useForm } from "../../hooks/useForm"
import { getHeroByName } from "../../selectors/getHeroByName";

import { HeroCard } from "../hero/HeroCard";

export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const [ formState , handleInputChange,] = useForm({
    searchText: q
  });
  
  const { searchText } = formState;
  const memoHeroes = useMemo(() => getHeroByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`)
    location.search = q;
  }

  return (
    <>
        <h1>Search a hero</h1>
        <hr />
        
        <div className="row">

          <div className="col-5">
            <h4>Formulario</h4>
            <hr />

            <form onSubmit={handleSearch}>
              <input
                type='text'
                placeholder="hero name..."
                autoComplete="off"
                className="form-control"
                name="searchText"
                value={searchText}
                onChange={handleInputChange}
              />
              <button 
                type="submit"
                className="btn btn-outline-info form-control mt-1"
              >
                Search...
              </button>
            </form>
          </div>

          <div className="col-7">
            <h4>Result</h4>
            <hr />
            {
              (q === '')
                ? <div className="alert alert-info">Buscar un heroe</div>
                : (memoHeroes.length === 0) && <div className="alert alert-danger">No hay resultados: { q }</div>
            }

            {
              memoHeroes.map(hero => (
                <HeroCard 
                  key={ hero.id }
                  { ...hero } 
                />
              ))
            }
          </div>

        </div>
    </>
  )
}
