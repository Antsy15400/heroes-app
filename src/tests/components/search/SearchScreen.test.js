import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom"
import { SearchScreen } from "../../../components/search/SearchScreen"

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))

describe('Pruebas en <SearchScreen />', () => {  

    test('Debe mostrarse correctamente con valores por defecto', () => {  
        const wrapper = mount( 
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        )
        expect( wrapper ).toMatchSnapshot()
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Buscar un heroe')
    })
    test('Debe mostrar a batman y el input con el valor del queryString', () => {  
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot()
        expect( wrapper.find('.alert-info').length ).toBe(0);
        expect( wrapper.find('.card-title').text().trim() ).toBe('Batman');
        expect( wrapper.find('input').prop('value') ).toBe('batman');
    })       
    test('Debe mostrar una alerta de que el heroe buscado no existe', () => {  
        const fakeHero = 'batman324234';
        const wrapper = mount(
            <MemoryRouter initialEntries={ [`/search?q=${fakeHero}`] }>
                <SearchScreen />
            </MemoryRouter>
        )
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-danger').exists() ).toBe(true);
        expect( wrapper.find('.alert-danger').text().trim() ).toBe(`No hay resultados: ${fakeHero}`)
    })
    test('El navigate debe modificar la url con los argumentos que se le pasen, mediante handleSeatch', () => {  

        const wrapper = mount(
            <MemoryRouter initialEntries={ [`/search`] }>
                <SearchScreen />
            </MemoryRouter>
        )
        wrapper.find('input').simulate('change',{
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').simulate('submit',{preventDefault(){}});
        
        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman')
    })

})