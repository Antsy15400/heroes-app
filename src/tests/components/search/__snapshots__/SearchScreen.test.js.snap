// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe mostrar a batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search a hero
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Formulario
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="hero name..."
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-info form-control mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card h-100"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4 "
            >
              <img
                alt="Batman"
                className="card-img"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> Debe mostrar una alerta de que el heroe buscado no existe 1`] = `
Array [
  <h1>
    Search a hero
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Formulario
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="hero name..."
          type="text"
          value="batman324234"
        />
        <button
          className="btn btn-outline-info form-control mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay resultados: 
        batman324234
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> Debe mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Search a hero
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Formulario
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="hero name..."
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-info form-control mt-1"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un heroe
      </div>
    </div>
  </div>,
]
`;
